---
title: "Cell_State_IDandAnnotation"
author: "Nathan Haywood"
date: "2024-10-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r}
library(Seurat) 
library(tidyselect)
library(dplyr)
library(reshape2)
library(DirichletReg)
library(data.table)
library(tibble)
library(circlize)
library(viridis)
library(tidyr)
library(ggplot2)
library(presto)
library(qs)
library(purrr)
library(UpSetR)
```

Include the code to run a quick Dirichlet regression to check against the cell state proportions
```{r}
dirichlet_regression = function(counts, covariates, formula){

  counts = as.data.frame(counts)
  counts$counts = DR_data(counts)
  data = cbind(counts, covariates)
  fit = DirichReg(counts ~ condition, data)
  
  # Get p-values
  u = summary(fit)
  pvals = u$coef.mat[grep('Intercept', rownames(u$coef.mat), invert=T), 4]
  v = names(pvals)
  pvals = matrix(pvals, ncol=length(u$varnames))
  rownames(pvals) = gsub('condition', '', v[1:nrow(pvals)])
  colnames(pvals) = u$varnames
  fit$pvals = pvals
  
  fit
}

matrix_barplot = function(data, group_by=NULL, pvals=NULL, xlab='', ylab='Frequency', value='mean', error='se', legend.title='Groups', colors='Paired', pos='dodge', border=NA,
                          out=NULL, nrow=1.5, ncol=1.5, coord_flip=FALSE, sig_only=F, do.facet=F){
  if(is.null(group_by)){group_by = rownames(data)}
  if(nlevels(group_by) == 0){group_by = as.factor(group_by)}
  
  # Select significant comparisons
  if(sig_only == TRUE){
    j = apply(pvals, 2, min) <= .05
    if(sum(j) == 0){return(NULL)}
    data = data[,j,drop=F]
    pvals = pvals[,j,drop=F]
  }
  
  # Construct input data
  names = colnames(data)
  data = data.frame(group=group_by, data)
  group_levels = levels(group_by)
  colnames(data)[2:ncol(data)] = names
  data = as.data.table(gather_(data, 'x', 'y', setdiff(colnames(data), 'group')))
  
  # Value function
  if(value == 'mean'){vf = mean} else if(value == 'median'){vf = median} else {stop()}
  
  # Error function
  se = function(x, na.rm=T){sd(x, na.rm=na.rm)/sqrt(length(x))}    
  if(error == 'sd'){ef = sd} else if(error == 'se'){ef = se} else {ef = function(x, ...){0}}
  
  # Estimate error bars
  data = data[,.(u=vf(y, na.rm=T), s=ef(y, na.rm=T)),.(group, x)]
  
  # Add p-values 1
  if(!is.null(pvals)){
    pvals = as.data.frame(pvals) %>% rownames_to_column('group') %>% gather(x, pval, -group) %>% as.data.table()
    setkeyv(data, c('x', 'group'))
    setkeyv(pvals, c('x', 'group'))
    data = merge(data, pvals, all=T)
    data$lab1 = ifelse(data$pval <= .001, '**', ifelse(data$pval <= .05, '*', ''))
  }
  
  if(coord_flip == TRUE){names = rev(names); group_levels=rev(group_levels)}
  data$x = factor(data$x, levels=names)    
  data$group = factor(data$group, levels=group_levels)
  
  # Get colors
  if(length(colors) == 1){colors = set.colors[1:length(group_levels)]}
  
  # Plot data
  if(pos == 'stack'){
    p = ggplot(data) + geom_bar(aes(x=x, y=u, fill=group), colour=border, size=.25, stat='identity')
    if(error %in% c('sd', 'se')){p = p + geom_errorbar(aes(x=x, ymin=u-s, ymax=u+s, fill=group), stat='identity', width=.25)}
  } else {
    pos = position_dodge(.9)
    p = ggplot(data) + geom_bar(aes(x=x, y=u, fill=group), colour=border, size=.25, stat='identity', position=pos)
    if(error %in% c('sd', 'se')){p = p + geom_errorbar(aes(x=x, ymin=u-s, ymax=u+s, fill=group), stat='identity', position=pos, width=.25)}
  }
  
  p = p + 
    scale_fill_manual(values=colors, name=legend.title) + xlab(xlab) + ylab(ylab) +
    scale_color_manual('', values=c('#000000', '#999999', '#cccccc'), guide='none')
  
  # Facet wrap
  if(do.facet == TRUE){
    p = p + facet_grid(group ~ ., scales='free')
  }
  
  dy = max(data$u + data$s, na.rm=T)*.01
  if(coord_flip == FALSE){
    p = p + theme(axis.text.x = element_text(angle = -45, hjust = 0))
    if(!is.null(pvals)){p = p + geom_text(aes(x=x, y=u+s+dy, label=lab1, group=group), hjust='center', vjust=0, size=5, angle=0, position=pos)}
  } else {
    p = p + coord_flip()
    if(!is.null(pvals)){p = p + geom_text(aes(x=x, y=u+s+dy, label=lab1, group=group), hjust='center', vjust=1, size=5, angle=90, position=pos)}
  }
  
  # Save plot
  if(!is.null(out)){save_plot(plot=p, filename=out, nrow=nrow, ncol=ncol)}
  p
}
```

Well focus primarily on the microglia here, but will also run a quick analysis of the other cell types.
```{r}
setwd("/Users/nate/Desktop/Mikes_ATF3_Data/Cell_Type_Objects")
microgliaObject <- qread("finalMicrogliaObject.qs")
astrocyteObject <- qread("finalAstrocyteObject.qs")
macrophageObject <- qread("finalMacrophageObject.qs")
oligoObject <- qread("finalOligoObject.qs")
OPCObject <- qread("finalOPCObject.qs")
```

In this cycle, well run the cluster ID call at a higher resolution to pull out some of the more subtle differences in the transcriptional profiles of the different states.
```{r}
microgliaObject <- NormalizeData(microgliaObject, normalization.method = "LogNormalize", scale.factor = 10000)
microgliaObject <- FindVariableFeatures(microgliaObject, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(microgliaObject)
microgliaObject <- ScaleData(microgliaObject, features = all.genes)
microgliaObject <- RunPCA(microgliaObject, features = VariableFeatures(object = microgliaObject))
microgliaObject<-FindNeighbors(microgliaObject, dims = 1:20)
microgliaObject<-FindClusters(microgliaObject, resolution = 0.5)   
microgliaObject<-RunUMAP(microgliaObject, dims = 1:20)
DimPlot(microgliaObject, reduction = "umap", label = T, pt.size = 0.02)
DimPlot(microgliaObject, reduction = "umap", group.by= "Genotype", label = T, pt.size = 0.02)
DimPlot(microgliaObject, group.by = "Genotype", split.by = "Sex", pt.size = 0.02, reduction = "umap", label = F)
micromarkers<-FindAllMarkers(microgliaObject, min.pct = 0.1) #not specifying the test type uses wilcox as a standard which is efficient by the presto package
```

Well use an upset plot to check that each of the clusters is being driven by gene expression differences and isn't the product of technical variates.
```{r}
m.markers<-mutate(micromarkers, Positive_gene=paste0(micromarkers$avg_logFC>0,"_", gene)) #Distinguish enriched and decreased DEGs
de_gs_by_con<-split(m.markers, m.markers$cluster)
de_gs_by_con <- map(de_gs_by_con, "Positive_gene") #Accounts for the sign of gene
#Run general upset plot, general and subtypes:
names(de_gs_by_con)
upset(fromList(de_gs_by_con), text.scale = 1.4, nsets=20, empty.intersections = "off", sets.bar.color = "#56B4E9")
```

Well run the dirichlet test here to take a quick look at which of the clusters are most directly driven by the genotype
```{r}
data<-microgliaObject@meta.data
#data <- data %>% subset(animal_id != "A")
data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)
cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")
g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals, value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```


These are marker genes from figure4 in Mikes paper that are helpful, but not perfect for trying to map the clusters between datasets
```{r}
FeaturePlot(microgliaObject, "Tmem119", pt.size = 0.01, label = T) #Mikes MG_0
FeaturePlot(microgliaObject, "P2ry12", pt.size = 0.01, label = T) #Mikes MG_0
FeaturePlot(microgliaObject, "Ctse", pt.size = 0.01, label = T) #Mikes MG_1
FeaturePlot(microgliaObject, "Sorl1", pt.size = 0.01, label = T) #Mikes MG_1
FeaturePlot(microgliaObject, "Cd74", pt.size = 0.01, label = T) #Mikes MG_2
FeaturePlot(microgliaObject, "Lpl", pt.size = 0.01, label = T) #Mikes MG_3
FeaturePlot(microgliaObject, "Ccl4", pt.size = 0.01, label = T) #Mikes MG_3
FeaturePlot(microgliaObject, "Lgals3", pt.size = 0.01, label = T) #Mikes MG_3
FeaturePlot(microgliaObject, "Atp6v0d2", pt.size = 0.01, label = T) #Mikes MG_4
FeaturePlot(microgliaObject, "Mki67", pt.size = 0.01, label = T) #Mikes MG_4
FeaturePlot(microgliaObject, "Gpnmb", pt.size = 0.01, label = T) #Mikes MG_4
FeaturePlot(microgliaObject, "Cenph", pt.size = 0.01, label = T) #Mikes MG_5
FeaturePlot(microgliaObject, "Adam33", pt.size = 0.01, label = T) #Mikes MG_7
FeaturePlot(microgliaObject, "Cxcl10", pt.size = 0.01, label = T) #Mikes MG_8
FeaturePlot(microgliaObject, "Ifit3", pt.size = 0.01, label = T) #Mikes MG_8
FeaturePlot(microgliaObject, "Oasl2", pt.size = 0.01, label = T) #Mikes MG_8

FeaturePlot(microgliaObject, "Atp6v1a", pt.size = 0.01, label = T) #Mikes MG_7
FeaturePlot(microgliaObject, "Ms4a7", pt.size = 0.01, label = T) #Mikes MG_7
FeaturePlot(microgliaObject, "Gpr137b", pt.size = 0.01, label = T) #Mikes MG_7
FeaturePlot(microgliaObject, "Egr2", pt.size = 0.01, label = T) #Mikes MG_3
FeaturePlot(microgliaObject, "Kcnj2", pt.size = 0.01, label = T) #Mikes MG_3
FeaturePlot(microgliaObject, "Efr3b", pt.size = 0.01, label = T) #Mikes MG_3
FeaturePlot(microgliaObject, "Atf3", pt.size = 0.01, label = T) #Mikes MG_3
```

Well quickly show some of the differences between the genotypes and how they impact the individual clusters
```{r}
DimPlot(microgliaObject, reduction = "umap", group.by= "Genotype",split.by = "seurat_clusters", label = T, pt.size = 0.02)
temp <- subset(microgliaObject, seurat_clusters %in% c("0", "3"))
DimPlot(temp, reduction = "umap", group.by= "Genotype",split.by = "seurat_clusters", label = T, pt.size = 0.02)
```

Well remove cluster 5 as a doublet cluster and rerun the analysis before doing DEG and a more robust DiffAbundance analysis
```{r}
microgliaObject <- subset(microgliaObject, seurat_clusters != "5")
microgliaObject <- NormalizeData(microgliaObject, normalization.method = "LogNormalize", scale.factor = 10000)
microgliaObject <- FindVariableFeatures(microgliaObject, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(microgliaObject)
microgliaObject <- ScaleData(microgliaObject, features = all.genes)
microgliaObject <- RunPCA(microgliaObject, features = VariableFeatures(object = microgliaObject))
microgliaObject<-FindNeighbors(microgliaObject, dims = 1:20)
microgliaObject<-FindClusters(microgliaObject, resolution = 0.5)   
microgliaObject<-RunUMAP(microgliaObject, dims = 1:20)
DimPlot(microgliaObject, reduction = "umap", label = T, pt.size = 0.02)
DimPlot(microgliaObject, reduction = "umap", group.by= "Genotype", label = T, pt.size = 0.02)
DimPlot(microgliaObject,group.by = "Genotype", split.by = "Sex", pt.size = 0.02, reduction = "umap", label = F)

micromarkers<-FindAllMarkers(microgliaObject, min.pct = 0.1) #not specifying the test type uses wilcox as a standard which is efficient by the presto package

m.markers<-mutate(micromarkers, Positive_gene=paste0(micromarkers$avg_logFC>0,"_", gene)) #Distinguish enriched and decreased DEGs
de_gs_by_con<-split(m.markers, m.markers$cluster)
de_gs_by_con <- map(de_gs_by_con, "Positive_gene") #Accounts for the sign of gene
#Run general upset plot, general and subtypes:
names(de_gs_by_con)
upset(fromList(de_gs_by_con), text.scale = 1.4, nsets=20, empty.intersections = "off", sets.bar.color = "#56B4E9")
```


```{r}
data<-microgliaObject@meta.data

data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)
cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")
g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals
                  ,   value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```

```{r}
data<-microgliaObject@meta.data

data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)

cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")


g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals
                  ,   value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```




```{r}
FeaturePlot(microgliaObject, "Tmem119", pt.size = 0.01, label = T) #Homeostatic
FeaturePlot(microgliaObject, "Lpl", pt.size = 0.01, label = T) #Dam1
FeaturePlot(microgliaObject, "Gpnmb", pt.size = 0.01, label = T) #Dam2
FeaturePlot(microgliaObject, "Cenph", pt.size = 0.01, label = T) #Proliferating
FeaturePlot(microgliaObject, "Ifit3", pt.size = 0.01, label = T) #Interferon
FeaturePlot(microgliaObject, "Cxcl10", pt.size = 0.01, label = T) #Interferon
FeaturePlot(microgliaObject, "Atf3", pt.size = 0.01, label = T)

microgliaObject$annotations <- NA

for (i in 1:length(microgliaObject$seurat_clusters)) {
  if(microgliaObject$seurat_clusters[i] == "1"){
    microgliaObject$annotations[i] <- "Homeostatic"
  }
  else if(microgliaObject$seurat_clusters[i] %in% c("2")){
    microgliaObject$annotations[i] <- "Lpl DAMs 1"
  }
  else if(microgliaObject$seurat_clusters[i] %in% c("3")){
    microgliaObject$annotations[i] <- "Lpl DAMs 2"
  }
  else if(microgliaObject$seurat_clusters[i] == "4"){
    microgliaObject$annotations[i] <- "Interferon"
  }
  else if(microgliaObject$seurat_clusters[i] == "5") {
    microgliaObject$annotations[i] <- "Gpnmb DAMs"
  }
  else if(microgliaObject$seurat_clusters[i] == "6"){
    microgliaObject$annotations[i] <- "Proliferating"
  }
  else{
    microgliaObject$annotations[i] <- "Undefined"
  }
}

setwd("/Users/nate/Desktop/Mikes_ATF3_Data/Cell_Type_Objects")
qsave(microgliaObject, "annotatedMicroglia.qs")
```

```{r}
temp <- microgliaObject
Idents(temp) <- 'seurat_cluster'
DimPlot(temp, reduction = "umap", label = T, pt.size = 0.02, repel = T, group.by = "annotations") + ggtitle ("Microglia Transcriptional States")
```

```{r}
for (i in 1:length(temp$seurat_clusters)) {
    if(temp$seurat_clusters[i] %in% c("2","3","5")){
        temp$annotations[i] <- "DAMs"
    }
}

```

```{r}



Idents(temp) <- 'annotations'
Cluster_Highlight_Plot(temp, "DAMs", color = "blue")
```



```{r}
astrocyteObject <- NormalizeData(astrocyteObject, normalization.method = "LogNormalize", scale.factor = 10000)
astrocyteObject <- FindVariableFeatures(astrocyteObject, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(astrocyteObject)
astrocyteObject <- ScaleData(astrocyteObject, features = all.genes)
astrocyteObject <- RunPCA(astrocyteObject, features = VariableFeatures(object = astrocyteObject))
astrocyteObject<-FindNeighbors(astrocyteObject, dims = 1:20)
astrocyteObject<-FindClusters(astrocyteObject, resolution = 0.2)   
astrocyteObject<-RunUMAP(astrocyteObject, dims = 1:20)
DimPlot(astrocyteObject, reduction = "umap", label = T, pt.size = 0.02)
DimPlot(astrocyteObject, reduction = "umap", group.by= "Genotype", label = T, pt.size = 0.02)
DimPlot(astrocyteObject,group.by = "Genotype", split.by = "Sex", pt.size = 0.02, reduction = "umap", label = F)
astromarkers<-FindAllMarkers(astrocyteObject, min.pct = 0.1) #not specifying the test type uses wilcox as a standard which is efficient by the presto package

m.markers<-mutate(astromarkers, Positive_gene=paste0(astromarkers$avg_logFC>0,"_", gene)) #Distinguish enriched and decreased DEGs
de_gs_by_con<-split(m.markers, m.markers$cluster)
de_gs_by_con <- map(de_gs_by_con, "Positive_gene") #Accounts for the sign of gene
#Run general upset plot, general and subtypes:
names(de_gs_by_con)
upset(fromList(de_gs_by_con), text.scale = 1.4, nsets=20, empty.intersections = "off", sets.bar.color = "#56B4E9")

data<-astrocyteObject@meta.data

data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)
cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")
g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals
                  ,   value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```

A couple of things in the data jump out that we want to take a closer look at. Most obvious is that clusters 4 and 6 are driven almost exclusively by the males. Its also worth taking a second look at cluster 7 to make sure it isnt contamination.
```{r}
DimPlot(astrocyteObject,group.by = "Genotype", split.by = "orig.ident", pt.size = 0.02, reduction = "umap", label = F)
temp <- subset(astrocyteObject, seurat_clusters %in% c("4", "6"))
DimPlot(temp,group.by = "Genotype", split.by = "orig.ident", pt.size = 0.02, reduction = "umap", label = F)

#View(astromarkers)
```
Cluster 7 appears to be a neuronal doublet cluster that well remove and rerun the seurat pipeline. Clusters 4 and 6 seem to be largely driven by 2 samples (1 case and 1 control) so well leave them for now and take a deeper look during the DEG analysis.

```{r}
astrocyteObject <- subset(astrocyteObject, seurat_clusters != "7")

astrocyteObject <- NormalizeData(astrocyteObject, normalization.method = "LogNormalize", scale.factor = 10000)
astrocyteObject <- FindVariableFeatures(astrocyteObject, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(astrocyteObject)
astrocyteObject <- ScaleData(astrocyteObject, features = all.genes)
astrocyteObject <- RunPCA(astrocyteObject, features = VariableFeatures(object = astrocyteObject))
astrocyteObject<-FindNeighbors(astrocyteObject, dims = 1:20)
astrocyteObject<-FindClusters(astrocyteObject, resolution = 0.2)   
astrocyteObject<-RunUMAP(astrocyteObject, dims = 1:20)
DimPlot(astrocyteObject, reduction = "umap", label = T, pt.size = 0.02)
DimPlot(astrocyteObject, reduction = "umap", group.by= "Genotype", label = T, pt.size = 0.02)
DimPlot(astrocyteObject,group.by = "Genotype", split.by = "Sex", pt.size = 0.02, reduction = "umap", label = F)
astromarkers<-FindAllMarkers(astrocyteObject, min.pct = 0.1) #not specifying the test type uses wilcox as a standard which is efficient by the presto package

m.markers<-mutate(astromarkers, Positive_gene=paste0(astromarkers$avg_logFC>0,"_", gene)) #Distinguish enriched and decreased DEGs
de_gs_by_con<-split(m.markers, m.markers$cluster)
de_gs_by_con <- map(de_gs_by_con, "Positive_gene") #Accounts for the sign of gene
#Run general upset plot, general and subtypes:
names(de_gs_by_con)
upset(fromList(de_gs_by_con), text.scale = 1.4, nsets=20, empty.intersections = "off", sets.bar.color = "#56B4E9")

data<-astrocyteObject@meta.data

data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)
cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")
g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals
                  ,   value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```

```{r}
macrophageObject <- NormalizeData(macrophageObject, normalization.method = "LogNormalize", scale.factor = 10000)
macrophageObject <- FindVariableFeatures(macrophageObject, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(macrophageObject)
macrophageObject <- ScaleData(macrophageObject, features = all.genes)
macrophageObject <- RunPCA(macrophageObject, features = VariableFeatures(object = macrophageObject))
macrophageObject<-FindNeighbors(macrophageObject, dims = 1:20)
macrophageObject<-FindClusters(macrophageObject, resolution = 0.6)   
macrophageObject<-RunUMAP(macrophageObject, dims = 1:20)
DimPlot(macrophageObject, reduction = "umap", label = T, pt.size = 0.02)
DimPlot(macrophageObject, reduction = "umap", group.by= "Genotype", label = T, pt.size = 0.02)
DimPlot(macrophageObject,group.by = "Genotype", split.by = "Sex", pt.size = 0.02, reduction = "umap", label = F)
macromarkers<-FindAllMarkers(macrophageObject, min.pct = 0.1) #not specifying the test type uses wilcox as a standard which is efficient by the presto package

m.markers<-mutate(macromarkers, Positive_gene=paste0(macromarkers$avg_logFC>0,"_", gene)) #Distinguish enriched and decreased DEGs
de_gs_by_con<-split(m.markers, m.markers$cluster)
de_gs_by_con <- map(de_gs_by_con, "Positive_gene") #Accounts for the sign of gene
#Run general upset plot, general and subtypes:
names(de_gs_by_con)
upset(fromList(de_gs_by_con), text.scale = 1.4, nsets=20, empty.intersections = "off", sets.bar.color = "#56B4E9")

data<-macrophageObject@meta.data

data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)
cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")
g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals
                  ,   value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```

```{r}
oligoObject <- NormalizeData(oligoObject, normalization.method = "LogNormalize", scale.factor = 10000)
oligoObject <- FindVariableFeatures(oligoObject, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(oligoObject)
oligoObject <- ScaleData(oligoObject, features = all.genes)
oligoObject <- RunPCA(oligoObject, features = VariableFeatures(object = oligoObject))
oligoObject<-FindNeighbors(oligoObject, dims = 1:20)
oligoObject<-FindClusters(oligoObject, resolution = 0.2)   
oligoObject<-RunUMAP(oligoObject, dims = 1:20)
DimPlot(oligoObject, reduction = "umap", label = T, pt.size = 0.02)
DimPlot(oligoObject, reduction = "umap", group.by= "Genotype", label = T, pt.size = 0.02)
DimPlot(oligoObject,group.by = "Genotype", split.by = "Sex", pt.size = 0.02, reduction = "umap", label = F)
oligomarkers<-FindAllMarkers(oligoObject, min.pct = 0.1) #not specifying the test type uses wilcox as a standard which is efficient by the presto package

m.markers<-mutate(oligomarkers, Positive_gene=paste0(oligomarkers$avg_logFC>0,"_", gene)) #Distinguish enriched and decreased DEGs
de_gs_by_con<-split(m.markers, m.markers$cluster)
de_gs_by_con <- map(de_gs_by_con, "Positive_gene") #Accounts for the sign of gene
#Run general upset plot, general and subtypes:
names(de_gs_by_con)
upset(fromList(de_gs_by_con), text.scale = 1.4, nsets=20, empty.intersections = "off", sets.bar.color = "#56B4E9")

data<-oligoObject@meta.data

data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)
cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")
g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals
                  ,   value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```




```{r}
OPCObject <- NormalizeData(OPCObject, normalization.method = "LogNormalize", scale.factor = 10000)
OPCObject <- FindVariableFeatures(OPCObject, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(OPCObject)
OPCObject <- ScaleData(OPCObject, features = all.genes)
OPCObject <- RunPCA(OPCObject, features = VariableFeatures(object = OPCObject))
OPCObject<-FindNeighbors(OPCObject, dims = 1:20)
OPCObject<-FindClusters(OPCObject, resolution = 0.6)   
OPCObject<-RunUMAP(OPCObject, dims = 1:20)
DimPlot(OPCObject, reduction = "umap", label = T, pt.size = 0.02)
DimPlot(OPCObject, reduction = "umap", group.by= "Genotype", label = T, pt.size = 0.02)
DimPlot(OPCObject,group.by = "Genotype", split.by = "Sex", pt.size = 0.02, reduction = "umap", label = F)
OPCmarkers<-FindAllMarkers(OPCObject, min.pct = 0.1) #not specifying the test type uses wilcox as a standard which is efficient by the presto package

m.markers<-mutate(OPCmarkers, Positive_gene=paste0(OPCmarkers$avg_logFC>0,"_", gene)) #Distinguish enriched and decreased DEGs
de_gs_by_con<-split(m.markers, m.markers$cluster)
de_gs_by_con <- map(de_gs_by_con, "Positive_gene") #Accounts for the sign of gene
#Run general upset plot, general and subtypes:
names(de_gs_by_con)
upset(fromList(de_gs_by_con), text.scale = 1.4, nsets=20, empty.intersections = "off", sets.bar.color = "#56B4E9")

data<-OPCObject@meta.data

data$Animal_Condition<-paste0(data$animal_id, "_" ,data$Genotype)
cell.freq<-as.matrix(as.data.frame.matrix(table(data$Animal_Condition, data$seurat_clusters)))
cell.pct = 100*cell.freq/rowSums(cell.freq)

meta.data<-data.frame(condition=sapply(strsplit(rownames(cell.freq), "_"), "[", 2))
rownames(meta.data)<-row.names(cell.freq)
meta.data #Make sure everything matches

cell.freq.pvals<-dirichlet_regression(counts=cell.freq, covariates=meta.data, formula=counts ~ condition)$pvals
#View(cell.freq.pvals) #Can write as csv for Supplementary information 
set.colors= c("grey", "#F8766D", "#7CAE00", "#00BFC4", "#C77CFF", "black", "yellow", "red")
g<-matrix_barplot(cell.pct, group_by=meta.data$condition, pvals=cell.freq.pvals
                  ,   value='mean',  error='se', legend.title='Condition')
g<-g+ylab(label = "Percent per sample")
g<-g+theme_classic()
g
```





















