---
title: "16_Pseudotime_Analysis"
author: "Nathan Haywood"
date: "2025-05-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(monocle3)
library(Seurat)
library(dplyr)
library(qs)
```


```{r}
setwd("/Users/nate/Desktop/Mikes_ATF3_Data/Cell_Type_Objects")
mg <- qread("3_annotatedMicroglia.qs")
DimPlot(mg, label = T, raster=FALSE)


expression_matrix <- mg[["RNA"]]$counts
cell_metadata <- mg[[]]
gene_annotation <- data.frame(
  gene_short_name = rownames(expression_matrix),
  row.names = rownames(expression_matrix)
)

# Create Monocle 3 CellDataSet
cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```


Ordering the cells in pseudotime
```{r}
cds <- preprocess_cds(cds, num_dim = 50)
cds <- reduce_dimension(cds)
cds <- cluster_cells(cds)
cds <- learn_graph(cds)
cds <- order_cells(cds)

plot_cells(cds)
plot_cells(cds, color_cells_by = "pseudotime")

plot_cells(cds, genes=c("Tmem119", "Lpl", "Gpnmb", "Ifit3", "Atf3"))

```




```{r}
#Run Monocle on OPCs in the demyelination condition (try it with both first)
#Using the filtered OPC dataset 

library(BiocManager)
library(monocle3)
library(Seurat)
library(tidyselect)
library(dplyr)
library(R.utils)
library(devtools)
#library(liger)
library(reshape2)
library(SingleCellExperiment)
library(qs)
library(SeuratWrappers)

#Load up the filtered, final OPC dataset: 
Opcs<-qread("/Users/nate/Desktop/Mikes_ATF3_Data/Cell_Type_Objects/3_annotatedMicroglia.qs")
DimPlot(Opcs)

#Subset OPCs by just deemyelinating lesion
Opcs.LPC<-Opcs

#Create the monocle object, phenodata is going to contain all the interesting metadata for these cells  
#Note this has changed significantly since Monocle 2
pd<-Opcs.LPC@meta.data
fd<-as.data.frame(row.names(as.matrix(GetAssayData(Opcs.LPC, slot = "counts"))))    #as.matrix(GetAssayData(object.2, slot = "counts")
row.names(fd)=fd$`row.names(as.matrix(GetAssayData(Opcs.LPC, slot = "counts"))`
colnames(fd)="gene_short_name"
object.2_matrix<-as.matrix(GetAssayData(Opcs.LPC, slot = "counts")) #Monocle likes to normalize itself.
cds<-new_cell_data_set(expression_data = as(object.2_matrix, "sparseMatrix"),
                       cell_metadata = pd,
                       gene_metadata = fd)

#Pull and manipulate the cell metadata
colData(cds)


#Monocle3 has preprocessing in single function. Default log normalizes, scales, PCA 
cds <- preprocess_cds(cds, num_dim = 50)

#NOTE THERE IS NO ALIGNMENT HERE 

## Step 2: Reduce the dimensions using UMAP
cds <- reduce_dimension(cds) #What does aligned mean in this context (preprocess method)

## Step 3: Cluster the cells
cds <- cluster_cells(cds)

## Step 4: Learn a graph
cds <- learn_graph(cds)
plot_cells(cds,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)

## Step 5: Order cells
cds <- order_cells(cds, reduction_method = "UMAP")
plot_cells(cds, show_trajectory_graph = F)

#Plot by different conditions and replicate
plot_cells(cds, color_cells_by="seurat_clusters")
plot_cells(cds, color_cells_by="animal_id")

#Plot some marker genes: 
marker_genes<- c( "Lpl", "Gpnmb", "Atf3")
plot_cells(cds,
           genes=marker_genes,
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)

marker_genes<- c( "Tmem119", "Ifit3", "Cenph")
plot_cells(cds,
           genes=marker_genes,
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)

#Plot the OPC states from Seurat analysis: 
plot_cells(cds, color_cells_by="seurat_clusters")
plot_cells(cds,
           color_cells_by = "seurat_clusters",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=TRUE,
           graph_label_size=10 , cell_size = 0.5, show_trajectory_graph=FALSE)

#Plot the OPCs in pseudotime 
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           show_trajectory_graph = FALSE,
           graph_label_size=1.5, cell_size = 0.5)

#THIS DOESN"T WORK 
#Calculate differential expression by pseudotime: 
#diff.exp.pseudo <- graph_test(cds, neighbor_graph="principal_graph")
#deg_filtered <- row.names(subset(diff.exp.pseudo, q_value < 0.05))

#Plot different marker genes as a function of pseudotime and either timepoint 
#or OPC cluster 
#Plot marker genes by pseudotime and ligerex cell-type 
#marker_genes <- c("Glis3", "Myo5b", "Col11a1", "Col4a6", "Ndst3", "Serpina3n", "Nhs", "Id3", "Dock7") #Longer list of good genes 
marker_genes <- c("Glis3", "Col4a6", "Serpina3n", "Mag") #Longer list of good genes 
marker_gene_cds <- cds[rowData(cds)$gene_short_name %in% marker_genes]

plot_genes_in_pseudotime(marker_gene_cds,
                         color_cells_by="seurat_clusters",
                         min_expr=0.5)
#plot_genes_in_pseudotime(marker_gene_cds,
#                         color_cells_by="Timepoint",
#                         min_expr=0.5)

```



```{r}
library(Seurat)
library(monocle3)
library(htmlwidgets)
library(qs)

### Reading in Seurat object

print("Readingin Seurat objects")

seurat <- qread("/Users/nate/Desktop/Mikes_ATF3_Data/Cell_Type_Objects/3_annotatedMicroglia.qs")

#seurat <- readRDS(file = input.dir)

### Re-dimension reduction for 3D rendering

#if (Dim = "3D") {
  
#  print ("Running UMAP 3D")
  
#  seurat <- RunUMAP(object = seurat, reduction = "pca", dims = 1:nPC, n.components = 3)
  
#  print("Clustering 3D")
  
#  seurat <- FindNeighbors(object=seurat, dims=1:nPC)
#  seurat <- FindClusters(object=seurat, resolution=cluster.res)
#  seurat[[sprintf("ClusterNames_%.1f_%dPC", cluster.res, nPC)]] <- Idents(object = seurat)
  
#}


### Building the necessary parts for a basic cds

# part one, gene annotations

gene_annotation <- as.data.frame(rownames(seurat@reductions[["pca"]]@feature.loadings), row.names = rownames(seurat@reductions[["pca"]]@feature.loadings))
colnames(gene_annotation) <- "gene_short_name"

# part two, cell information

#cell_metadata <- as.data.frame(seurat@assays[["RNA"]]@counts@Dimnames[[2]], row.names = #seurat@assays[["RNA"]]@counts@Dimnames[[2]])
cell_metadata <- seurat@meta.data
colnames(cell_metadata) <- "barcode"

# part three, counts sparse matrix

#object.2_matrix<-as.matrix(GetAssayData(seurat, slot = "counts"))

New_matrix <- as.matrix(GetAssayData(seurat, slot = "counts"))
New_matrix <- New_matrix[rownames(seurat@reductions[["pca"]]@feature.loadings), ]
expression_matrix <- New_matrix


### Construct the basic cds object

cds_from_seurat <- new_cell_data_set(expression_matrix,
                                     cell_metadata = cell_metadata,
                                     gene_metadata = gene_annotation)


### Construct and assign the made up partition

recreate.partition <- c(rep(1, length(cds_from_seurat@colData@rownames)))
names(recreate.partition) <- cds_from_seurat@colData@rownames
recreate.partition <- as.factor(recreate.partition)

cds_from_seurat@clusters@listData[["UMAP"]][["partitions"]] <- recreate.partition


### Assign the cluster info

list_cluster <- seurat@meta.data[[sprintf("seurat_clusters", 0.5, 20)]]
names(list_cluster) <- seurat@assays[["RNA"]]@layers[["data"]]@Dimnames[[2]]

cds_from_seurat@clusters@listData[["UMAP"]][["clusters"]] <- list_cluster


### Could be a space-holder, but essentially fills out louvain parameters

cds_from_seurat@clusters@listData[["UMAP"]][["louvain_res"]] <- "NA"


### Assign UMAP coordinate

#cds_from_seurat@reduce_dim_aux@listData[["UMAP"]] <-seurat@reductions[["umap"]]@cell.embeddings
cds_from_seurat@reduce_dim_aux@listData[["UMAP"]] <- Embeddings(seurat, reduction = "umap")

### Assign feature loading for downstream module analysis

#cds_from_seurat@preprocess_aux$gene_loadings <- seurat@reductions[["pca"]]@feature.loadings
cds_from_seurat@reduce_dim_aux$gene_loadings <- Loadings(seurat, reduction = "umap")

### Learn graph, this step usually takes a significant period of time for larger samples

print("Learning graph, which can take a while depends on the sample")

cds_from_seurat <- learn_graph(cds_from_seurat, use_partition = T)

plot_cells(cds, color_cells_by="seurat_clusters")
```






```{r}
library(Seurat)
library(SeuratWrappers)
library(monocle3)
library(htmlwidgets)
library(qs)

### Reading in Seurat object

print("Reading in Seurat objects")

seurat <- qread("/Users/nate/Desktop/Mikes_ATF3_Data/Cell_Type_Objects/3_annotatedMicroglia.qs")

DimPlot(seurat, label = TRUE)

seurat <- UpdateSeuratObject(seurat)
cds <- as.cell_data_set(seurat)
cds <- cluster_cells(cds)
cds <- learn_graph(cds)
plot_cells(cds, label_groups_by_cluster = TRUE, label_leaves = FALSE, label_branch_points = FALSE)

plot_cells(cds, color_cells_by="seurat_clusters")
plot_cells(cds, color_cells_by="animal_id")

cds <- order_cells(cds)
plot_cells(cds, color_cells_by = "pseudotime")
```








```{r}
library(monocle3)
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(ggplot2)
library(patchwork)
library(magrittr)


InstallData("hcabm40k")
data("hcabm40k")

hcabm40k <- UpdateSeuratObject(hcabm40k)
hcabm40k <- SplitObject(hcabm40k, split.by = "orig.ident")
for (i in seq_along(hcabm40k)) {
    hcabm40k[[i]] <- NormalizeData(hcabm40k[[i]]) %>% FindVariableFeatures()
}
features <- SelectIntegrationFeatures(hcabm40k)
for (i in seq_along(along.with = hcabm40k)) {
    hcabm40k[[i]] <- ScaleData(hcabm40k[[i]], features = features) %>% RunPCA(features = features)
}


anchors <- FindIntegrationAnchors(hcabm40k, reference = c(1, 2), reduction = "rpca", dims = 1:30)
integrated <- IntegrateData(anchors, dims = 1:30)


integrated <- ScaleData(integrated)
integrated <- RunPCA(integrated)
integrated <- RunUMAP(integrated, dims = 1:30, reduction.name = "UMAP")
integrated <- FindNeighbors(integrated, dims = 1:30)
integrated <- FindClusters(integrated)
DimPlot(integrated, group.by = c("orig.ident", "ident"))


cds <- as.cell_data_set(integrated)
cds <- cluster_cells(cds)
p1 <- plot_cells(cds, show_trajectory_graph = FALSE)
p2 <- plot_cells(cds, color_cells_by = "partition", show_trajectory_graph = FALSE)
wrap_plots(p1, p2)


integrated.sub <- subset(as.Seurat(cds, assay = NULL), monocle3_partitions == 1)
cds <- as.cell_data_set(integrated.sub)
cds <- learn_graph(cds)
plot_cells(cds, label_groups_by_cluster = FALSE, label_leaves = FALSE, label_branch_points = FALSE)

max.avp <- which.max(unlist(FetchData(integrated.sub, "AVP")))
max.avp <- colnames(integrated.sub)[max.avp]
cds <- order_cells(cds, root_cells = max.avp)
plot_cells(cds, color_cells_by = "pseudotime", label_cell_groups = FALSE, label_leaves = FALSE, 
    label_branch_points = FALSE)
```





```{r}

mg <- AddMetaData(mg, pseudotime(cds), col.name = 'pseudotime')
head(mg)


selected <- subset(mg, mg$pseudotime != "Inf")

temp <- data.frame(matrix(ncol = 5, nrow = 0))
names <- c("animal", "state", "pseudotime_ave", "Genotype", "ClusterID")
colnames(temp) <- names
for (i in unique(selected$animal_id)) {
  animal <- subset(selected, selected$animal_id == i)
  for(j in unique(animal$annotations)){
    state <- subset(animal, animal$annotations == j)
    apst <- mean(state$pseudotime)
    temp <- rbind(temp, list(i, j, apst, unique(state$Genotype), unique(state$seurat_clusters)))
  }
}
colnames(temp) <- names
View(temp)



for(k in unique(temp$ClusterID)){
  t2 <- subset(temp, temp$ClusterID == k)
  print(t2 |> 
    tidyplot(x = Genotype, y = pseudotime_ave, color = Genotype) |> 
    add_sem_errorbar() |> 
    add_mean_dot() |>
    view_plot(title = k)
  )

}

temp |> 
  tidyplot(x = Genotype, y = pseudotime_ave, color = Genotype) |> 
  add_sem_errorbar() |> 
  add_mean_dot() 

t_test_result <- t.test(pseudotime_ave ~ Genotype, data = temp)
print(t_test_result)
```













